docs/design/node_agent_config-design.md

```diff
@@ -0,0 +1,91 @@
+# NodeAgentConfig configuration with restic/kopia
+Date: 2023-07-25
+
+## Abstract
+
+New configuration for the file system backup & restore that will be used by the OADP and allow to choose Restic or Kopia uploader type.
+
+## Background
+
+For the file system backup and restore the Velero may use Restic or Kopia as the uploader mechanism. A number of tests were [performed](https://velero.io/docs/main/performance-guidance/) by the Velero community to compare those mechanisms. In many cases Kopia uploader is a much better performing mechanism and as such should be added to the OADP operator.
+
+## Goals
+
+- A new option `nodeagentconfig` to allow configuration of restic or kopia uploader
+- Backwards compatibility with the current OADP configuration schema options, namely `restic`
+- Preparation for the future deprecation of the current `restic` configuration option (from OADP 1.4+)
+- Allow new schema to be used by the datamover node agent
+
+## Non Goals
+- Enablment of datamover node agent
+- Deprecation of the "restic" configuration option
+- E2E tests for the `kopia` or `restic` uplaoder
+
+## High-Level Design
+
+Since new `nodeagentconfig` configuration option is a sibling of the `restic` one, the `restic` will be used as part of the new `nodeagentconfig` with addition of one `UploaderType` option that will be either `kopia` or `restic`.
+
+Without that option a default `restic` will be used. That default option may change in the future.
+
+## Detailed Design
+
+A new structure will be added, that includes the ResticConfig inline and extends this with one new parameter `UploaderType`:
+
+```go
+type NodeAgentConfig struct {
+	// Embedding ResticConfig
+	// +kubebuilder:validation:Optional
+	ResticConfig `json:",inline"`
+
+	// The type of uploader to transfer the data of pod volumes, the supported values are 'restic' or 'kopia'
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Enum=restic;kopia
+	UploaderType string `json:"uploaderType,omitempty"`
+}
+```
+
+The above `NodeAgentConfig` is a member of `ApplicationConfig`, which already includes `ResticConfig`, however we do not replace the `ResticConfig`, instead for backwards compatibility we add new `NodeAgentConfig` parameter:
+
+```go
+// ApplicationConfig defines the configuration for the Data Protection Application
+type ApplicationConfig struct {
+	Velero *VeleroConfig `json:"velero,omitempty"`
+	// (deprecation warning) ResticConfig is the configuration for restic server.
+	// Restic is for backwards compatibility and will be replaced by nodeagentconfig
+	// with the OADP 1.4+
+	// +optional
+	Restic *ResticConfig `json:"restic,omitempty"`
```

Comment:
Is there validation, such that a user can not set Restic and NodeAgent?

Reply:
@mpryc @shawn-hurley We should definitely allow only one struct to be included. Either Restic *or* NodeAgent must be empty. Validation should fail if both are included, and validation should give a deprecation warning if Restic is included.

Reply:
+100 on deprecation warning pretty sure there is a new header you can set that will do that. 

I wonder if we are going to be able to, unless we have a webhook though.

Reply:
We could have "Reconciled with warnings" status?

Reply:
The fields is there and I will use it:
   https://book.kubebuilder.io/reference/markers/crd.html

It was added in thtat PR:
https://github.com/kubernetes-sigs/controller-tools/blob/ecaffbc2dc09cd5e09c07b692d899dbbc8fdda0e/pkg/crd/markers/crd.go#L52

Reply:
My understanding is that the above `// +kubebuilder:deprecatedversion:warning=<string>` is not working with field validations and just CRD's and their deprecation. If that's correct we can't really use it here as we are not deprecating CRD, just one field within already existing one.

Reply:
Just to confirm: we are renaming `ResticConfig` to `NodeAgentCommonFields`? should we change here as well?